### ISSUE TITLE
[docman] Document the parameterizable behavior

### ISSUE THREAD

**Issue by stevenrombauts** [joomlatools/developer.joomlatools.com#14]
See : http://www.joomlatools.com/membercenter/forums/topic/2947
See : http://www.joomlatools.com/support/forums/topic/4043
See : http://www.joomlatools.com/support/forums/topic/4382-customs-text-field-for-a-category

Explain how people can add custom fields using template overrides.

Markdown available: 
Framework: https://docs.google.com/document/d/1UMpQYXzb3Uu7ISghaIl8S9Mcq0f19ihqp6-SYZDbL1M/edit
Docman Extension : https://docs.google.com/document/d/1L-A7HYYcQlUFXufHbLFhmT3iUBbrAFsdFCj4RBCm354/edit

Resources :

- http://mysql.rjweb.org/doc.php/eav
- http://www.slideshare.net/MindfireSolutions/eav-data-model-concepts
- https://en.wikipedia.org/wiki/Entity%E2%80%93attribute%E2%80%93value_model

**Comment by stevenrombauts** [comment]
re#14 first run at custom fields

Branch: feature/14-parameterizable-tutorial
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/23172831b2f49817c7c7275275d16c1620f9fc6b|joomlatools-documentation:23172831b2]]

**Comment by stevenrombauts** [comment]
@johanjanssens 

I took a stab at this. I used a format that is close to what I laid out in http://guides.nooku.org/contribute/write-a-tutorial.html . Its an experiment. 
I like the fact that the background is at the bottom and I've added a "Relevant Resources" section as well. The reader can choose how far they want to go, but the solution is right at the top.

**Comment by stevenrombauts** [comment]
Good work. Feedback :

1. I would move the background information to framework/behaviors where you can explain this behavior in detail. The info on how to use a custom column for this is really advanced info and I would put that as a final part. 'Using a custom column'.

2. I would add additional info on how to use the validation javascript behavior to add validation to the field being added. This would allow the field to be validatable. HTML5 already has build in validation support for various input types, adding additional links to that info would be helpful.

**Comment by stevenrombauts** [comment]
@johanjanssens 

Great. Will do all that.

**Comment by stevenrombauts** [comment]
re#14 - breakup background into its own framework level piece

Branch: feature/14-parameterizable-tutorial
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/9fda6ee7c2e8a4d939e8a8cc0fa5445e74a358ab|joomlatools-documentation:9fda6ee7c2]]

**Comment by stevenrombauts** [comment]
re#14 - refining some of the major points, moving the examples around

Branch: feature/14-parameterizable-tutorial
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/b2ef242ae8386df1e4d77c7b7f6a5bca6c087689|joomlatools-documentation:b2ef242ae8]]

**Comment by stevenrombauts** [comment]
re#14 - refining some of the major points

Branch: feature/14-parameterizable-tutorial
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/7b623b583b0fc45002306edfb9982d2ab1704aca|joomlatools-documentation:7b623b583b]]

**Comment by stevenrombauts** [comment]
@cameronbarr Here is an example of how this tutorial could be used : http://www.joomlatools.com/support/forums/topic/4382-customs-text-field-for-a-category

**Comment by stevenrombauts** [comment]
@johanjanssens I've got it. 

I've been fleshing out the database portion of the behaviors to support a more detailed explanation of the parameterizable db behavior. Its rough but available at https://github.com/nooku/guides.nooku.org/blob/a68046be7482a544fc98eea06cb34f3437b85451/behaviors.md . The intention is to copy to the framework portion of developer.joomlatools.com as well. 

I'll make the mentioned additions to the core docman tutorial piece and try to finish the framework pieces (with some input from you) in the coming weeks.

**Comment by stevenrombauts** [comment]
@cameronbarr This is good work, you are writing an introduction into behaviors. I would consider this as part of the Nooku documentation, we do not need it in Joomlatools, can refer to Nooku if needed.

However, the goal of this ticket and for Joomlatools is to demonstrate and explain how the parameterizable behavior works. The question to answer :  'How can you add a custom field to DOCman ?' 

The parameterizable behavior allows you to add fields and save data for them, per document or per category on the fly, that field and data can then be used on the frontend. Both can be done using a simple template override. 

The goal of this tutorial is to explain this in basic terms and demonstrate it with an example. For example, this  http://www.joomlatools.com/support/forums/topic/4382-customs-text-field-for-a-category is question from a customer this tutorial should be able to answer. 

I would add a section Custom Fields in the DOCman documentation to add this tutorial into. Again we are not trying to explain the framework to our customers, just teaching them how to use it to do basic things.

**Comment by stevenrombauts** [comment]
@johanjanssens 

Of course you are right about the point of this ticket and where that document I showed you should go. Its why its in the guides repo. I was simply letting you know what's happening in connection with this (#14)  ticket. 
What I am trying to do all at once is write complementary pieces that go from DOCman to Framework (to Guide). My hope is that by doing that they knowledge related will be consistent and connected. 
I'll update  this [[url:https://github.com/joomlatools/developer.joomlatools.com/blob/feature/14-parameterizable-tutorial/extensions/docman/parameterizable.md|parameterizable piece]] with the comments from [[url:https://timble.assembla.com/spaces/joomlatools-documentation/tickets/14--docman--document-the-parameterizable-behavior?comment=661077463#comment:661077463|here]], i.e. html5 validation, etc.

Re: Custom Fields section...*check*

**Comment by stevenrombauts** [comment]
@cameronbarr Makes total sense. Just wanted to let you know that from a Joomlatools perspective the most important part is how to use parameterizalble using an template override. For DOCman specific would call this "Custom Fields".  Keep up the good work :)

**Comment by stevenrombauts** [comment]
@johanjanssens Ok, Great.

**Comment by stevenrombauts** [comment]
re #14 - reworked and reviewed

Branch: feature/14-parameterizable-tutorial
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/25d1afcff8ddf7af2adce88b44037eab0b9e8caa|joomlatools-documentation:25d1afcff8]]

**Comment by stevenrombauts** [comment]
Hi @johanjanssens 
I am assigning these two pieces for your review. They were originally written in md format but have them in GDocs for you to comment. 

The "Custom Fields" piece will go into the extensions/docman folder and the Parameterizable behavior piece can go in both the framework/ folder of djt.com and into the guides (with the other behavioral pieces that are to be reviewed).

**Comment by stevenrombauts** [comment]
@cameronbarr Can you please share the second document with me ? I do not have access to it on Google Docs. Thanks.

**Comment by stevenrombauts** [comment]
@cameronbarr Did further review on the parameterizable document. Still not 100% happy with is, making progress. I also noticed that the urls i added to the document at the workshop are no longer in there ? Do you still have those ? I would like to add them as references at the bottom as they explain the EAV principle quite well and based on that include a little EAV intro.

**Comment by stevenrombauts** [comment]
@johanjanssens, thanks. 
I think we need some clarity on the EAV thing. I interpret Parameterizable as an alternative method to EAV modelling and not to be under the umbrella "EAV". The references you provide suggest the same. Especially, http://mysql.rjweb.org/doc.php/eav which would make Parameterizable as part of the "A Solution" part of that page. 

So, if we are to mention EAV at all, the approach should be to show that storing json in a column for a dynamic schema differs from the sometimes more familiar EAV model. Otherwise I think it may be a little out of scope for what we are trying to achieve with this piece.  

I'm willing to admit I may be wrong here, but that's my take atm.

**Comment by stevenrombauts** [comment]
re #14 - Refactored new content to Jekyll

Branch: feature/14-parameterizable-tutorial
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/971c4e3b6af79730b3f1b2b86210338c32114d1a|joomlatools-documentation:971c4e3b6a]]

**Comment by stevenrombauts** [comment]
@cameronbarr Oki, thanks, lets get the Custom Fields documentation for DOCman published first, once that is ready we will finalise the framework part on the parameterizable behaviour itself.  I still want to include a little bit of background information about EAV. 

From : https://en.wikipedia.org/wiki/Entity%E2%80%93attribute%E2%80%93value_model

"Entity–attribute–value model (EAV) is a data model to describe entities where the number of attributes (properties, parameters) that can be used to describe them is potentially vast, but the number that will actually apply to a given entity is relatively modest. In mathematics, this model is known as a sparse matrix. EAV is also known as object–attribute–value model, vertical database model and open schema."

**Comment by stevenrombauts** [comment]
@cameronbarr Added additional info to the DOCman/Joomlatools tutorial; lets focus on that one first. 

Goal : I would like this tutorial to explain how to :

1. add a ‘price field’ to a document form (in the site and administrator)
2. display the price information on the frontend (in the document list and details view) (edited)

Use the Hierarchical List as the base. 

Does that help you to know what you need to work further on ?

**Comment by stevenrombauts** [comment]
@johanjanssens 
Getting closer on the Custom Fields piece: 
https://docs.google.com/document/d/1L-A7HYYcQlUFXufHbLFhmT3iUBbrAFsdFCj4RBCm354/edit

I've rewritten the pieces, and been more explicit with the file copy steps. I'll ask you to read it over quickly and tell me if we're going the right way. I'm committing in the repo as well as the GDoc.

**Comment by stevenrombauts** [comment]
re#14 - including list and backend edit use cases

Branch: feature/14-parameterizable-tutorial
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/e3cb1eaa57849072c5ea882de0cdf2a4dcf13cc8|joomlatools-documentation:e3cb1eaa57]]

**Comment by stevenrombauts** [comment]
@cameronbarr Yes, I like this much better . It's to the point, flows well and get's the user to where he needs to be. I added some smaller suggestions. All yours.

**Comment by stevenrombauts** [comment]
@johanjanssens great. Will take it.

**Comment by stevenrombauts** [comment]
re#14 - add conclusion and incorporate suggested edits

Branch: feature/14-parameterizable-tutorial
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/fc620a8b0c6464de6e6a843ed72bf72ea44bf7e7|joomlatools-documentation:fc620a8b0c]]

**Comment by stevenrombauts** [comment]
@johanjanssens, I've written the conclusion for this, and will do a screen shot of the front end as a last step. Then I think we can mark this complete. We can leave out the reference to the framework paramterizable behavior description until its done.

**Comment by stevenrombauts** [comment]
@cameronbarr I agree, left you a few more notes, screenshot and this is ready indeed.

**Comment by stevenrombauts** [comment]
@johanjanssens, this is live.

**Comment by stevenrombauts** [comment]
Check the links in the article, I get some 404's.

Little feedback on the semantics:

* Remove the <strong> markup from the headings
* Remove the blockquotes, seems like blockquotes have been used to create visual difference, if this is needed we need to create some extra CSS classes you can use, ping me to discuss this further

**Comment by stevenrombauts** [comment]
re#14 first run at custom fields

Branch: master
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/23172831b2f49817c7c7275275d16c1620f9fc6b|joomlatools-documentation:23172831b2]]

**Comment by stevenrombauts** [comment]
re#14 - breakup background into its own framework level piece

Branch: master
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/9fda6ee7c2e8a4d939e8a8cc0fa5445e74a358ab|joomlatools-documentation:9fda6ee7c2]]

**Comment by stevenrombauts** [comment]
re#14 - add screenshot to repo

Branch: master
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/52a7a6915748f8c07e300b75a65d4e8b8b855284|joomlatools-documentation:52a7a69157]]

**Comment by stevenrombauts** [comment]
re#14 - refining some of the major points

Branch: master
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/7b623b583b0fc45002306edfb9982d2ab1704aca|joomlatools-documentation:7b623b583b]]

**Comment by stevenrombauts** [comment]
re #14 - reworked and reviewed

Branch: master
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/25d1afcff8ddf7af2adce88b44037eab0b9e8caa|joomlatools-documentation:25d1afcff8]]

**Comment by stevenrombauts** [comment]
re#14 - add screenshot to page

Branch: master
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/1b6c36b76a170b75b5098b00309fa7597a4250e2|joomlatools-documentation:1b6c36b76a]]

**Comment by stevenrombauts** [comment]
re#14 - refining some of the major points, moving the examples around

Branch: master
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/b2ef242ae8386df1e4d77c7b7f6a5bca6c087689|joomlatools-documentation:b2ef242ae8]]

**Comment by stevenrombauts** [comment]
re #14 - Content improvements

# Conflicts:
#	extensions/docman/custom-parameters.md

Branch: feature/14-parameterizable-tutorial
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/d2ee3c73946a4fcfac5d1ecce120d2a5d7d7c3f8|joomlatools-documentation:d2ee3c7394]]

**Comment by stevenrombauts** [comment]
I fixed all the mentioned issues.

**Comment by stevenrombauts** [comment]
re#14 - re incorporating these files here

Branch: feature/34-dispatcher-events
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/11a1f3d20635e89cc33eb9a276ce5d97bc9a50a3|joomlatools-documentation:11a1f3d206]]

**Comment by stevenrombauts** [comment]
re#14 - remove from this branch

Branch: staging
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/b57c75d0de7e162b5f9b087fe4714ba9b87e82c0|joomlatools-documentation:b57c75d0de]]

**Comment by stevenrombauts** [comment]
re#14 - add code highligting

Branch: staging
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/af25a79ea8bbb6161de55fc4e072fd0e22391afe|joomlatools-documentation:af25a79ea8]]

**Comment by stevenrombauts** [comment]
re#14 - spacing

Branch: staging
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/cbecd8312b9c8162f485e80d34492651f230ada4|joomlatools-documentation:cbecd8312b]]

**Comment by stevenrombauts** [comment]
re#14 - last edit

Branch: staging
Commit: [[url:https://github.com/joomlatools/developer.joomlatools.com/commit/d66d390c7d7867df61b54f53a9130b15fa378448|joomlatools-documentation:d66d390c7d]]
